kotlinx-coroutines-core/common/src/channels/Channel.kt

```diff
@@ -85,7 +85,23 @@ public interface SendChannel<in E> {
      * then it calls `onUndeliveredElement` before throwing an exception.
      * See "Undelivered elements" section in [Channel] documentation for details on handling undelivered elements.
      */
-    public fun offer(element: E): Boolean
+    public fun offer(element: E): Boolean {
+        val result = trySend(element)
+        if (result.isSuccess) return true
+        throw recoverStackTrace(result.exceptionOrNull() ?: return false)
```

Comment:
Here we have a minor design problem: such implementation of `offer` does not invoke `onUndeliveredElement`.

I've added a workaround 3d44ab93074508c70f9eea67414a3ea0c0213a5c that we can maintain until 1.7 version (where `offer` will become `HIDDEN` function with a default implementation).